### PR TITLE
Make commit-msg-hook.js pay more attention to the nohook option

### PR DESCRIPTION
**Pull request by all3fox**
If the nohook is true in either package.json or .npmrc:
 - do not install the hook (setup.js)
 - do not run the hook (if installed before) (commit-msg-hook.js)

Put index.js as ./node_modules/.bin/commitplease because it does not
care about the nohook option and will always run if triggered
manually.

This is the first step to solving on-demand validation #70. Next I am thinking of modifying index.js to make it recognise commit ranges specified from console (like grunt-commitplease)

The problem is the duplication of code in `setup.js` and `commit-msg-hook.js`. It can be solved if I make `index.js` "not executable" at `require(commitplease)` and make it export a couple of functions (like reading settings from different places or running commitplease) instead.

We can make this PR a work-in-progress one so that I could show what I mean? However, this stage is complete and could be used right away :)

### CODE REVIEW

**Comment by all3fox** [comment]
Things to know for this PR:

1. `process.cwd()` is set to point at the root of repository when git triggers its hooks: [stackoverflow][1] (and a [link to git docs][2] from there).
2. `uninstall.js` that uses 3rd-party modules is buggered in `npm@3.x`, see npm/npm#11894 

And I could not find any reliable way to let commitplease know if it is being installed locally or globally. Do you perhaps know how? Let's assume it is not possible to reliably know. So, I made commitplease behave exactly the same in any case:

`install.js`:
1. create a `.git/hooks` directory two levels up of `process.cwd()`
2. copy hook to `.git/hooks/commit-msg` if there was not one before

`uninstall.js`
1. look for a `commit-msg` two levels up in `.git/hooks`
2. if that is exactly our `commit-msg-hook.js`, remove it

When this scenario is triggered for a global install, it is just extra actions. On the downside, an empty `.git/hooks` will be left upon uninstall.

When this scenarios is triggered for a local install, it works similar to as before this PR.

Update: will be fixing CI problems now

[1]: http://stackoverflow.com/a/37927943/1269892
[2]: https://git-scm.com/docs/githooks/2.9.0

**Comment by jzaefferer** [comment]
I haven't tested it and probably didn't dig deep enough on the diff, overall its looking good though.

Now that the nohook option applies at runtime (not just during install), we should update the documentation for it.

**Comment by all3fox** [comment]
@jzaefferer I have finished nohook, global install + CLI and updated the README accordingly. PR is finally ready for review :)

File: README.md [hunk 1]
```diff
@@ -32,7 +39,24 @@ The following ways to begin a commit message are special and always valid:
...
+### Global install
+
+Navigate to your repository and run the global commitplease executable. By default, it will check all the commit messages. Other examples include (just anything you can pass to `git log` really):
+
+| Use case | command |
+|----------|---------|
+| Check all commits on branch `master` | `commitplease master` |
+| Check the latest 1 commit | `commitplease -1` |
+| Check all commits between `84991d` and `2021ce` | `commitplease 84991d..2021ce` |
+| Check all commits starting with `84991d` | `commitplease 84991d..` |
```
Comment (jzaefferer) [inline README.md]: I think a common usecase would be to check all commits on a non-master branch, like a feature branch, but ignore any commits that are already on master.

File: README.md [hunk 2]
```diff
@@ -200,4 +224,5 @@ Support this project by [donating on Gratipay](https://gratipay.com/jzaefferer/)
 
 [1]: http://contribute.jquery.org/commits-and-pull-requests/#commit-guidelines
 [2]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit
-[3]: https://github.com/typicode/husky
\ No newline at end of file
+[3]: https://github.com/typicode/husky
+[4]: bit.ly/commit-range
```
Comment (jzaefferer) [inline README.md]: This should include a protocol. I'd link to https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection#Commit-Ranges directly.

File: index.js [hunk 3]
```diff
@@ -1,23 +1,37 @@
 var fs = require('fs')
+var ini = require('ini')
 var path = require('path')
-var root = path.resolve(__dirname, '../..')
 var chalk = require('chalk')
+var Gtools = require('git-tools')
```
Comment (jzaefferer) [inline index.js]: How about `git` as the variable name?

File: index.js [hunk 4]
```diff
@@ -1,23 +1,37 @@
...
+  var pkg = path.join(process.cwd(), 'package.json')
+  var npm = path.join(process.cwd(), '.npmrc')
+
+  pkg = fs.existsSync(pkg) && require(pkg) || {}
+  npm = fs.existsSync(npm) && ini.parse(fs.readFileSync(npm, 'utf8')) || {}
+
+  pkg = pkg.commitplease || {}
+  npm = npm.commitplease || {}
+
+  return merge(pkg, npm)
```
Comment (jzaefferer) [inline index.js]: Can we use `Object.assign`? Would be able to drop the `mout` dependency then.

**Comment by jzaefferer** [comment]
A few minor nits. The CLI looks good to me.

**Comment by all3fox** [comment]
Thanks, fixed:

1. `standard` insists that a constructo has to be capitalized, so `Git` instead of `git` for `git-tools`
1. Dropped `mout` entirely, using `Object.assign` everywhere now
1. Added docs on the "check only new commits on feature branch"
1. Fixed `https` for links

please review :)

**Comment by all3fox** [comment]
Oh, looks like `node 0.12` does not have `Object.assign` and I have not dropped `mout` from `test.js`

Will be fixing CI failure now

**Comment by all3fox** [comment]
@jzaefferer I have updated `test.js` to drop `mout`. Now `node 4` and `node 6` work but `node 0.12` does not. It does not have `Object.assign`

Do we go back to `mout` or drop `node 0.12`? Any other idea?

**Comment by jzaefferer** [comment]
I don't think that's worth dropping node 0.12 for. Let's try a more specialised package for now, something like https://www.npmjs.com/package/object-assign - I'd actually prefer a polyfill, so that we can keep using `Object.assign`.

**Comment by jzaefferer** [comment]
Everything else is looking good!

**Pull request merged**